backend/src/apiserver/storage/pipeline_store_kubernetes.go

```diff
@@ -0,0 +1,502 @@
+	k8sPipelineVersion, err := k.getK8sPipelineVersion(pipelineVersionId)
+	if err != nil {
+		if strings.Contains(err.Error(), "ResourceNotFoundError:") {
+			return nil
+		}
+	}
+
+	err = k.client.Delete(context.TODO(), k8sPipelineVersion)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return util.NewInternalServerError(err, "Failed to delete the pipeline version")
+	}
+
+	return nil
+}
+
+func (k *PipelineStoreKubernetes) getK8sPipeline(pipelineId string) (*v2beta1.Pipeline, error) {
+	// TODO: Consider making the ID namespace/name format to make querying efficient
+	pipelines := v2beta1.PipelineList{}
+
+	listOptions := []ctrlclient.ListOption{ctrlclient.UnsafeDisableDeepCopy}
+
+	if !common.IsMultiUserMode() && common.GetPodNamespace() != "" {
+		listOptions = append(listOptions, ctrlclient.InNamespace(common.GetPodNamespace()))
+	}
+
+	// Be careful, the deep copy is disabled here to reduce memory allocations
+	err := k.client.List(context.TODO(), &pipelines, listOptions...)
+	if err != nil {
+		return nil, util.NewInternalServerError(err, "Failed to find the pipeline")
+	}
+
+	for _, k8sPipeline := range pipelines.Items {
+		if string(k8sPipeline.UID) == pipelineId {
+			return &k8sPipeline, nil
+		}
+	}
+
+	return nil, util.NewResourceNotFoundError("Pipeline", pipelineId)
+}
+
+func (k *PipelineStoreKubernetes) getK8sPipelineVersions() (*v2beta1.PipelineVersionList, error) {
+	// TODO: Consider making the ID namespace/name format to make querying efficient
+	pipelineVersions := v2beta1.PipelineVersionList{}
+
+	listOptions := []ctrlclient.ListOption{ctrlclient.UnsafeDisableDeepCopy}
+
+	if !common.IsMultiUserMode() && common.GetPodNamespace() != "" {
+		listOptions = append(listOptions, ctrlclient.InNamespace(common.GetPodNamespace()))
+	}
+
+	// Be careful, the deep copy is disabled here to reduce memory allocations
+	err := k.client.List(context.TODO(), &pipelineVersions, listOptions...)
+	if err != nil {
+		return nil, util.NewInternalServerError(
+			err, "Failed to find the pipeline version associated with this pipeline version",
+		)
+	}
+
+	return &pipelineVersions, nil
+}
+
+func (k *PipelineStoreKubernetes) getK8sPipelineVersion(pipelineVersionId string) (*v2beta1.PipelineVersion, error) {
+	pipelineVersions, err := k.getK8sPipelineVersions()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, k8sPipelineVersion := range pipelineVersions.Items {
+		if string(k8sPipelineVersion.UID) == pipelineVersionId {
+			return &k8sPipelineVersion, nil
+		}
+	}
+
+	return nil, util.NewResourceNotFoundError("PipelineVersion", pipelineVersionId)
+}
+
+func (k *PipelineStoreKubernetes) createPipelineVersionWithPipeline(pipeline *model.Pipeline, pipelineVersion *model.PipelineVersion) (*model.PipelineVersion, error) {
+	if pipelineVersion.PipelineSpecURI != "" {
+		return nil, util.NewBadRequestError(ErrUnsupportedField, "The pipeline spec URI field is not supported")
```

Comment:
it is not immediately clear why this is unsupported, can we get an inline code comment for why this is the case

Reply:
I think it could be supported. I think it makes pipelines not immutable anymore since you can change the contents of the URI without changing the URI, but if your pipeline is huge and etcd rejects it, this might be the only valid workaround.

Reply:
@rimolive what did you decide here? It seems like it may just work if we allow it to be set.